### DISCUSSION TITLE
Serving large files locally

### DISCUSSION THREAD

**Discussion by keeely** [mitmproxy/mitmproxy#6534] [General]
I'm seeing some problems serving large local files.  I think there's some point beyond which mitmproxy doesn't work very well in this respect.  Anything up to about 50MB is mostly working, once I go over that it falls over.  I wondered if this is a known issue and is there any work-around?

```
$ cat config.yaml
map_local:
  - "|~m GET|fake.com|random_text.txt"
```

Create a large file full of random data, e.g.

```
openssl rand -base64 $((120 * 1024 * 1024)) > random_text.txt
```

Run mitmproxy:

```
mitmproxy --set confdir=.
```
Run Chrome:

```
"/Applications/Google Chrome.app/Contents/MacOS/Google Chrome" --proxy-server=localhost:8080
```

Go to fake.com in the browser.  With this filesize it's failing almost all of the time, and I get the 'snap' Chrome screen..  Works consistently below about 50MB.

**Comment by keeely** [comment]
PS: In case it's relevant:
Mitmproxy: 10.1.5
Python:    3.10.13
OpenSSL:   OpenSSL 3.1.4 24 Oct 2023
Platform:  macOS-14.1.2-arm64-arm-64bit

**Comment by mhils** [comment]
I can reproduce this, but it's not a mitmproxy bug! 😄 

```shell
$ cat response-headers.txt
HTTP/1.1 200 OK
Server: mitmproxy 11.0.0.dev
Content-Type: text/plain
content-length: 170393600


$ unix2dos response-headers.txt  # ensure \r\n
$ cat response-headers.txt random_text.txt > full.txt
$ cat full.txt | nc -l 8080
```

still crashes Chrome.

**Comment by keeely** [comment]
I think it is a mitmproxy bug, although maybe not one that can easily be fixed.  One of the reasons for proxying is that you fake responses, so if the remote site serves a XXMB file and you want to replace it, then that's what you have to serve locally, you don't really have a choice in it.  Unfortunately this wasn't an example I concocted for fun.  I do actually need to serve a >100MB javascript file locally, sadly such things exist out there!

**Reply by mhils** [reply]
See my example above that reproduces Chrome's behavior without mitmproxy at all, just with netcat. I don't see how you arrive at the conclusion that this is a mitmproxy bug.

**Reply by keeely** [reply]
OK fair enough.  At first I thought you were comparing apples to oranges, then I tried it with a 'proper' web server (Nginx) and it does the same thing.  I'm now wondering if my mistake is displaying the file in the browser.  Probably loading it with JS api call will involve less overhead.  Never mind.
